### ISSUE TITLE
epoch taken by python-specfile is null, in srpm header is 0

### ISSUE THREAD

**Issue by praiskup** [fedora-copr/copr#2848]
As pointed out in https://github.com/fedora-copr/copr/issues/2839

```
Running SRPMResults tool
Package info: {
    "name": "python-copr",
    "epoch": null,
    "version": "1.129",
    "release": "1.git.68.9e71174",
    "exclusivearch": [],
    "excludearch": []
}
SRPMResults finished
```

```Running RPMResults tool
Package info:
{
    "packages": [
        {
            "name": "python-copr-doc",
            "epoch": 0,
            "version": "1.129",
            "release": "1.git.68.9e71174.fc39",
            "arch": "noarch"
        },
...
```

Those RPM data are taken from RPM headers; but if taken from spec-file, Epoch is not `null` but `0`. Is this a problem actually? Shouldn't we migrate Epoch from `null` to `0`?

**Comment by praiskup** [comment]
I mean - in case of `specfile` failure, we fallback to reading this data from srpm header, which likely would provide `0`.

**Comment by praiskup** [comment]
Also note that specfile returns string, while rpm header returns number:

```
Running SRPMResults tool
Package info: {
    "name": "tar",
    "epoch": "2",
...
```
Running RPMResults tool
Package info:
{
    "packages": [
        {
            "name": "tar-debugsource",
            "epoch": 2,
```

**Comment by praiskup** [comment]
It seems that we should stay as close to RPM behavior as possible, and work with integers or None.  WDYT?

@nforro JFYI, don't you want to return integer value for "epoch" in specfile automatically?

**Comment by nforro** [comment]
> @nforro JFYI, don't you want to return integer value for "epoch" in specfile automatically?

I see your point, but `epoch` represents the value of the `Epoch` tag, which can contain macros for example, so it has to be a string. If it is `None`, it means there is no `Epoch` tag in the spec file. `expanded_epoch` could be an integer, but still, I think it should follow the same logic and return `None` if there is no `Epoch` tag.

On a related note, be aware that there could be multiple `Epoch` tags in a spec file, and you may not always get what you want. For example here: https://src.fedoraproject.org/rpms/samba/blob/9602a07a36b5fbe33eca0c55c883b1a163ac2f64/f/samba.spec#_207
If `fedora` is not defined, you would expect epoch to be 0, but `spec.epoch` (and `spec.expanded_epoch`) will give you 2, because it takes the first tag and ignores the rest. You can get both values by iterating over tags, but that's probably not what you are after. I think would be safer to use `spec.rpm_spec.sourceHeader.epoch` (which also gives you an integer). Or use `spec.tags(spec.parsed_sections.package).content.epoch.expanded_value`.

**Comment by praiskup** [comment]
Wow, what a topic :-) 

The `epoch` and `expanded_epoch` is then something we don't want to use, I think. And then return values of specfile in such case are not our business :-) (but yes, returning `None` is IMO a valid approach, because that's what `rpm` returns for unspecified `Epoch:`, too).

Well, we probably want to use `spec.rpm_spec.sourceHeader.epoch`, but I'm not sure about the runtime consequences. Let's see.

**Comment by nforro** [comment]
To explain a bit more what the differences are:
`spec.epoch` is a just a convenience shortcut for `spec.tags(spec.sections.package).content.epoch.value`. You operate on an unprocessed spec file and all parsing is done by the `specfile` library. The `expanded_` variant is the same value postprocessed with `rpm` to expand macros.
When you replace `spec.sections.package` with `spec.parsed_sections.package`, you operate on a spec file processed by `rpm`, with all conditions resolved and all macros expanded (which means `.value` and `.expanded_value` are equivalent). Further parsing is still done by the `specfile` library.
`spec.rpm_spec` gives you the underlying `rpm.spec` object that represents the spec file parsed by `rpm`.

**Comment by praiskup** [comment]
Thank you for all the explanations, going to parse your comments once more tomorrow :-) because I'm already too tired and dumb now.  But obviously, the current approach [doesn't work well](https://download.copr-dev.fedorainfracloud.org/results/praiskup/test-2023-08-09/srpm-builds/02914729/builder-live.log):
```
Running SRPMResults tool
Package info: {
    "name": "texlive",
    "epoch": 11,
    "version": "%{tl_version}",
    "release": "69",
    "exclusivearch": [
        "aarch64",
        "ppc64le",
        "s390x",
        "x86_64",
        "noarch"
    ],
    "excludearch": []
}
SRPMResults finished
```

**Comment by praiskup** [comment]
Similar issue breaking the behave test:
```
                  "source_package": {
                      "name": "python-copr-common",                                    
                      "url": "https://download.copr-dev.fedorainfracloud.org/results/praiskup/test-that-autospec-macros-are-expanded-1691653215.3202565/srpm-builds/02914830/python-copr-common-0.13.1-3.src.rpm",                            
                      "version": "0.13.1-%autorelease"
                  },              
```

**Comment by nforro** [comment]
Right, at the very least, you should use `expanded_version` etc.

**Comment by praiskup** [comment]
Sorry, closed by mistake by a commit message.  @FrostyX promised to take a look at the remaining things.

**Comment by praiskup** [comment]
Should be fixed by 76ade5dce9c6e057a4e3e9f98c5986b4cff44718
